rest-server/templates/src/main/scala/com/actionml/templates/navhinting/NavHintingAlgorithm.scala

```diff
@@ -0,0 +1,219 @@
+        params = p.algorithm.copy()
+        // init the in-memory model, from which predicitons will be made and to which new conversion Journeys will be added
+        dataset.activeJourneysDAO.find(allCollectionObjects).foreach { j =>
+
+          activeJourneys += (j._id -> j.trail)
+        }
+        // TODO: create model from conversion Trails
+        // trigger Train and wait for finish, then input will trigger Train and Query will us the latest trained model
+        model = dataset.navHintsDAO.findOne(allCollectionObjects).getOrElse(Hints(Map.empty)).hints
+        //trainer = Some(new NavHintTrainer(convertedJourneys, params, resourceId, model, this).self)
+        // TODO: should train and wait for completion
+        Valid(true)
+      } else { //bad decay function name
+        Invalid(ParseError(s"Bad decayFunction: ${p.algorithm.decayFunction}"))
+      }
+    }
+  }
+
+  override def input(datum: NavHintingAlgoInput): Validated[ValidateError, Boolean] = {
+      logger.trace(s"Train Nav Hinting Model with datum: $datum")
+      //trainer.get ! Train(datum)
+      val activeJourney = activeJourneys.get(datum.event.entityId)
+      val converted = datum.event.properties.conversion.getOrElse(false)
+      if (converted) { // update the model with the active journeys and remove it from active
+        if (activeJourney.nonEmpty) { // have an active journey
+          updateModel(Journey(datum.event.entityId, activeJourney.get), DateTime.parse(datum.event.eventTime))
+          activeJourneys -= datum.event.entityId // remove once converted
+        } else { // new event from this user, start a journey
+          activeJourneys += datum.event.entityId -> Seq((datum.event.targetEntityId, DateTime.parse(datum.event.eventTime)))
+        }
+      } else { // no conversion so just update activeJourney
+        if (activeJourney.nonEmpty) { // have an active journey so update
+          activeJourneys += (datum.event.entityId -> updateTrail(
+            datum.event.targetEntityId,
+            DateTime.parse(datum.event.eventTime),
+            activeJourney.get))
+        } else { // no conversion, no journey, create a new one
+          activeJourneys += datum.event.entityId -> Seq((datum.event.targetEntityId, DateTime.parse(datum.event.eventTime)))
+        }
+      }
+      Valid(true)
+  }
+
+  /** add the event to the end of an active journey subject to length limits */
+  def updateTrail(navId: String, timeStamp: DateTime, trail: Seq[(String, DateTime)]): Seq[(String, DateTime)] = {
+    val newTrail = trail :+ (navId, timeStamp)
+    newTrail.takeRight(params.numQueueEvents.getOrElse(50))
+  }
+
+
+  /** update the model with a Future for every input. This may cause Futures to accumulate */
+  def updateModel(convertedJourney: Journey,  now: DateTime): Unit = {
+    applyDecayFunction(convertedJourney, now).map { weightedVectors =>
```

Comment:
this maps a Future

Reply:
Do we have the possibility to have overlapping Futures here, both modifying the `var model`?

Reply:
summarizing slack conversation: we can combine these two futures and return resulting future to the upper level. e.g. like that:
`def updateModel(...): Future[Unit] = {
    applyDecayFunction.map( <merge operation> ).flatMap(<write model file>)`

Reply:
done, it will show up in the next PR